llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -2427,6 +2470,15 @@ SDValue RISCVTargetLowering::LowerCall(CallLoweringInfo &CLI,
     InVals.push_back(RetValue);
   }
 
+  // Load X31 from the stack if function call overwrites X31(which is used
+  // as a register to store pointer authentication code) for Ibex pointer
+  // authentication.
+  if (shouldSaveOrLoadPAC(MF)) {
+    SDValue Val = DAG.getLoad(XLenVT, DL, Chain, PACSpillSlot,
+                              MachinePointerInfo::getFixedStack(MF, PACFI));
+    Chain = DAG.getCopyToReg(Chain, DL, RISCV::X31, Val);
```

Comment:
I *think* that we might need to glue this CopyToReg with the other bits here (in addition to the chain). This should help with ensuring that the copytoreg is always in the right place, if i understand correctly, which might solve our runtime library calls issue.

Reply:
I added `Glue = Chain.getValue(1)`; after getCopyToReg, but when I test it I get the following error:
```
llc: ../include/llvm/CodeGen/SelectionDAGNodes.h:1138: llvm::SDValue::SDValue(llvm::SDNode *, unsigned int): Assertion `(!Node || !ResNo || ResNo < Node->getNumValues()) && "Invalid result number for the given node!"' failed.
```

Reply:
That error means you are trying to get an element from the result "array" that is out of bounds. That's because there is no glue at that index. You don't get a glue value because you also didn't pass a glue argument (after `Val`). You can check the debug output. You'll find lines that look like this:

```
t19: ch,glue = CopyToReg t18, Register:i32 $x10, t16
```

That one includes the glue as a second element of the result, so there you can use `.getValue(1)`. In other circumstances (e.g. `i32,ch,glue`) you'd use different indices.
I may be wrong about this, but I *think* it might not be possible to just add the glue to the existing code. Basically, it seems that we need all of the things to be properly glued, and the load doesn't support glue, so it breaks the entire thing. I think one option might be to replace the load with a custom Selection DAG node that represents the same load but supports the glue, and that would be custom lowered to the load later.

Reply:
Yeah Luis is looking into this to help you too. He said he came across something similar.

Reply:
@mmxsrup Yesterday I managed to glue everything, including the load, but that didn't solve the issue. I'm afraid I won't be able to solve this in time for the end of the GSoC.

Reply:
@luismarques Thank you for your help! I will try to fix this bug on another occasion.